packages/sdk/src/filter/filter.ts

```diff
@@ -1,297 +1,259 @@
    *
-   * @hidden
+   * @param decoders - Single decoder or array of decoders to unsubscribe from. All decoders must share the same pubsubTopic.
+   * @returns Promise that resolves to true if unsubscription was successful, false otherwise.
    *
-   * @param decoders The decoders to use for the subscription.
-   * @param callback The callback function to use for the subscription.
-   * @param opts Optional protocol options for the subscription.
+   * @example
+   * // Unsubscribe from a single decoder
+   * await filter.unsubscribe(decoder);
    *
-   * @returns A Promise that resolves to a function that unsubscribes from the subscription.
+   * @example
+   * // Unsubscribe from multiple decoders at once
+   * await filter.unsubscribe([decoder1, decoder2]);
    *
-   * @remarks
-   * This method should not be used directly.
-   * Instead, use `createSubscription` to create a new subscription.
+   * @example
+   * // Handle unsubscription failure
+   * const success = await filter.unsubscribe(decoder);
+   * if (!success) {
+   *   console.error("Failed to unsubscribe");
+   * }
    */
-  public async subscribeWithUnsubscribe<T extends IDecodedMessage>(
-    decoders: IDecoder<T> | IDecoder<T>[],
-    callback: Callback<T>
-  ): Promise<Unsubscribe> {
-    const uniquePubsubTopics = this.getUniquePubsubTopics<T>(decoders);
+  public async unsubscribe<T extends IDecodedMessage>(
+    decoder: IDecoder<T> | IDecoder<T>[]
+  ): Promise<boolean> {
+    const decoders = Array.isArray(decoder) ? decoder : [decoder];
 
-    if (uniquePubsubTopics.length === 0) {
-      throw Error(
-        "Failed to subscribe: no pubsubTopic found on decoders provided."
-      );
+    if (decoders.length === 0) {
+      throw Error("Cannot unsubscribe with 0 decoders.");
     }
 
-    if (uniquePubsubTopics.length > 1) {
+    const pubsubTopics = decoders.map((v) => v.pubsubTopic);
+    const contentTopics = decoders.map((v) => v.contentTopic);
+
+    // doing this for simplicity, we can enable unsubscribing with more than one PubsubTopic at once later when requested
+    if (!this.isSamePubsubTopic(decoders)) {
       throw Error(
-        "Failed to subscribe: all decoders should have the same pubsub topic. Use createSubscription to be more agile."
+        `Cannot unsubscribe with more than one pubsub topic at the same time, got pubsubTopics:${pubsubTopics}`
       );
     }
 
-    const { subscription, error } = await this.createSubscription(
-      uniquePubsubTopics[0]
+    log.info(
+      `Unsubscribing from content topic: ${contentTopics}, pubsub topic: ${pubsubTopics}`
     );
 
-    if (error) {
-      throw Error(`Failed to create subscription: ${error}`);
+    const supportedPubsubTopic = this.connectionManager.pubsubTopics.includes(
+      pubsubTopics[0]
+    );
+    if (!supportedPubsubTopic) {
+      throw Error(
+        `Pubsub topic ${pubsubTopics[0]} has not been configured on this instance.`
+      );
+    }
```

Comment:
DRY, I think all this logic is duplicated in `subscribe` function

Reply:
I am not really fun on `throwIf` functions and I believe it is a bad code smell if it is used on a big code base
but it seems reasonable within one class - added `priv` functions for it